### PR TITLE
Support default value for Byte column

### PR DESCRIPTION
**Pull request by jackjlli**
This PR supports default value for Byte column.

Unit tests added.

### CODE REVIEW

File: pinot-common/src/test/java/org/apache/pinot/common/data/FieldSpecTest.java [hunk 1]
```diff
@@ -117,6 +117,17 @@ public void testFieldSpec() {
     Assert.assertEquals(fieldSpec1.toString(), fieldSpec2.toString());
     Assert.assertEquals(fieldSpec1.hashCode(), fieldSpec2.hashCode());
     Assert.assertEquals(fieldSpec1.getDefaultNullValue(), 1L);
+
+    // Metric field with default null value for byte column.
+    fieldSpec1 = new MetricFieldSpec();
+    fieldSpec1.setName("btyeMetric");
```
Comment (siddharthteotia) [inline pinot-common/src/test/java/org/apache/pinot/common/data/FieldSpecTest.java]: typo?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java [hunk 2]
```diff
@@ -303,6 +304,10 @@ protected void createColumnV1Indices(String column)
         dictionaryElementSize = StringUtil.encodeUtf8(stringDefaultValue).length;
         sortedArray = new String[]{stringDefaultValue};
         break;
+      case BYTES:
+        Preconditions.checkState(defaultValue instanceof byte[]);
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java]: We should probably check if the default value byte array's length should be >= 1?

File: pinot-common/src/test/java/org/apache/pinot/common/data/FieldSpecTest.java (hunk 1 above)
Comment (jackjlli) [reply pinot-common/src/test/java/org/apache/pinot/common/data/FieldSpecTest.java]: good catch 👍

File: pinot-core/src/test/resources/data/newColumnsSchema3.json [hunk 3]
```diff
@@ -16,6 +16,11 @@
     {
       "dataType": "DOUBLE",
       "name": "newDoubleMetric"
+    },
+    {
+      "dataType": "BYTES",
+      "defaultNullValue": "00000008000000ac0000",
```
Comment (kishoreg) [inline pinot-core/src/test/resources/data/newColumnsSchema3.json]: is this hexstring of the byte array

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java [hunk 4]
```diff
@@ -303,6 +304,10 @@ protected void createColumnV1Indices(String column)
         dictionaryElementSize = StringUtil.encodeUtf8(stringDefaultValue).length;
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java]: we should use something that is independent of the encoding on the system. The file might be created/saved in a different encoding right

File: pinot-core/src/test/resources/data/newColumnsSchema3.json (hunk 3 above)
Comment (jackjlli) [reply pinot-core/src/test/resources/data/newColumnsSchema3.json]: This is one value of hyperLogLog

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java [hunk 5]
```diff
@@ -303,6 +304,10 @@ protected void createColumnV1Indices(String column)
...
         sortedArray = new String[]{stringDefaultValue};
         break;
+      case BYTES:
+        Preconditions.checkState(defaultValue instanceof byte[]);
+        sortedArray = new ByteArray[] {new ByteArray((byte[]) defaultValue)};
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java]: The defaultValue is coming from JSON schema, correct? The expectation is to specify ` Hex.encodeHexString(bytes)`  in the schema. Which means that this code would need to use `Hex.decodeHex` here.

Also, if we don't have this documented, let's also add the expectation on how to specify the default value in the schema for BYTES.

File: pinot-core/src/test/resources/data/newColumnsSchema3.json (hunk 3 above)
Comment (jackjlli) [reply pinot-core/src/test/resources/data/newColumnsSchema3.json]: Yes, this is helix string of the byte array.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java (hunk 5 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java]: Actually the value is already of byte array type. We can document what/how to put a default value for byte column.

**Comment by jackjlli** [comment]
@mayankshriv tests for both HLL and TDigest columns added.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 6]
```diff
@@ -574,8 +575,14 @@ public static void addColumnMetadataInfo(PropertiesConfiguration properties, Str
     if (defaultNullValue == null) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: I don't think defaultNullValue can ever be null

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java [hunk 7]
```diff
@@ -574,8 +575,14 @@ public static void addColumnMetadataInfo(PropertiesConfiguration properties, Str
...
       defaultNullValue = fieldSpec.getDefaultNullValue();
     }
-    properties.setProperty(V1Constants.MetadataKeys.Column.getKeyFor(column, DEFAULT_NULL_VALUE),
-        String.valueOf(defaultNullValue));
+    if (fieldSpec.getDataType() == FieldSpec.DataType.BYTES && defaultNullValue instanceof byte[]) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: We just need the second check

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 6 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: Correct, we have default null value for all kinds of types in FieldSpec. 
Removed.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java (hunk 7 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/SegmentColumnarIndexCreator.java]: Removed.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java]: @kishoreg We always encode String using UTF-8. As long as the files are generated with Pinot library, the String will be encoded with UTF-8. Record reader should be able to decode other format if necessary.

**Pull request merged**